cmd/util/ledger/util/payload_grouping.go

```diff
@@ -0,0 +1,252 @@
+
+	return &PayloadAccountGroup{
+		Address:  address,
+		Payloads: g.payloads[accountStartIndex:accountEndIndex],
+	}, nil
+}
+
+// Len returns the number of accounts
+func (g *PayloadAccountGrouping) Len() int {
+	return len(g.indexes)
+}
+
+// GroupPayloadsByAccount takes a list of payloads and groups them by account.
+// it uses nWorkers to sort the payloads by address and find the start and end indexes of
+// each account.
+func GroupPayloadsByAccount(
+	log zerolog.Logger,
+	payloads []*ledger.Payload,
+	nWorkers int,
+) *PayloadAccountGrouping {
+	if len(payloads) == 0 {
+		return &PayloadAccountGrouping{}
+	}
+	p := sortablePayloads(payloads)
+
+	start := time.Now()
+	log.Info().
+		Int("payloads", len(payloads)).
+		Int("workers", nWorkers).
+		Msg("Sorting payloads by address")
+
+	// sort the payloads by address
+	sortPayloads(0, len(p), p, make(sortablePayloads, len(p)), nWorkers)
+	end := time.Now()
+
+	log.Info().
+		Int("payloads", len(payloads)).
+		Str("duration", end.Sub(start).Round(1*time.Second).String()).
+		Msg("Sorted. Finding account boundaries in sorted payloads")
+
+	start = time.Now()
+	// find the indexes of the payloads that start a new account
+	indexes := make([]int, 0, len(p))
+	for i := 0; i < len(p); {
+		indexes = append(indexes, i)
+		i = p.FindLastOfTheSameKey(i) + 1
+	}
+	end = time.Now()
+
+	log.Info().
+		Int("accounts", len(indexes)).
+		Str("duration", end.Sub(start).Round(1*time.Second).String()).
+		Msg("Done grouping payloads by account")
+
+	return &PayloadAccountGrouping{
+		payloads: p,
+		indexes:  indexes,
+	}
+}
+
+// payloadToAddress takes a payload and return:
+// - (address, true, nil) if the payload is for an account, the account address is returned
+// - ("", false, nil) if the payload is not for an account
+// - ("", false, err) if running into any exception
+// The zero address is used for global Payloads and is not an actual account
+func payloadToAddress(p *ledger.Payload) (common.Address, error) {
+	k, err := p.Key()
+	if err != nil {
+		return common.ZeroAddress, fmt.Errorf("could not find key for payload: %w", err)
+	}
+
+	id, err := convert.LedgerKeyToRegisterID(k)
+	if err != nil {
+		return common.ZeroAddress, fmt.Errorf("error converting key to register ID")
+	}
+
+	if len([]byte(id.Owner)) != flow.AddressLength {
+		return common.ZeroAddress, nil
+	}
+
+	address, err := common.BytesToAddress([]byte(id.Owner))
+	if err != nil {
+		return common.ZeroAddress, fmt.Errorf("invalid account address: %w", err)
+	}
+
+	return address, nil
+}
+
+type sortablePayloads []*ledger.Payload
+
+func (s sortablePayloads) Len() int {
+	return len(s)
+}
+
+func (s sortablePayloads) Less(i, j int) bool {
+	return s.Compare(i, j) < 0
+}
+
+func (s sortablePayloads) Compare(i, j int) int {
+	// sort descending to force one of the big accounts to be more at the beginning
+	return bytes.Compare(
+		s[j].EncodedKey()[:encodedKeyAddressPrefixLength],
+		s[i].EncodedKey()[:encodedKeyAddressPrefixLength],
```

Comment:
the non-account payloads are indeed shorter. 
when sorting the payloads the non account registers might get dispersed amongst the other accounts.

However this is not a problem, since all non-account payloads are skipped later, so they get filtered out.

Reply:
Oh I see, thanks for clarifying. :+1:   Could we add a comment for this?